### ISSUE TITLE
Different results for shapenet

### ISSUE THREAD

**Issue by bellenliu** [yizhiwang96/Slice3D#7]
Hi,

Thank you for your excellent work on the Slice3D project! I followed the instructions in the "Regression-based Slicing" sections of your README file. However, I obtained different results compared to what is shown in your paper. Specifically, the shape appears uncompleted. I attached screenshot of my results below.

Could you please let me know if there is anything I might be doing wrong? Thank you!

![Image](https://github.com/user-attachments/assets/430f1987-e772-47da-a058-74124babf8b2)

<img width="694" alt="Image" src="https://github.com/user-attachments/assets/152795e0-c095-424d-b847-bb353fb971b1" />

**Comment by yizhiwang96** [comment]
Thank you for your interest in our work!

From the figure you provided, it seems that you are using **axis-aligned slicing**, which needs camera parameters to perform 3D to 2D projection. Did you use the camera parameters to train the model? and use estimated ones to test the model?

**Comment by bellenliu** [comment]
Thank you for your reply. Is this the **CamerNet** that perform 3D to 2D projection? Is it to train this **CamerNet**  and then add the --est_campose parameter in python reconstruct.py? 

Could you please provide the **CamerNet**   training code and let me know if there is anything else needed? Thank you!


![Image](https://github.com/user-attachments/assets/e311dc3c-7ce4-4abd-a510-09343484ae07)

**Comment by bellenliu** [comment]
In addition, could you please provide the eval code for calculating the evaluation index?  Thank you!

**Comment by ChongminWang** [comment]
> Hi,
> 
> Thank you for your excellent work on the Slice3D project! I followed the instructions in the "Regression-based Slicing" sections of your README file. However, I obtained different results compared to what is shown in your paper. Specifically, the shape appears uncompleted. I attached screenshot of my results below.
> 
> Could you please let me know if there is anything I might be doing wrong? Thank you!
> 
> ![Image](https://github.com/user-attachments/assets/430f1987-e772-47da-a058-74124babf8b2)
> 
> <img alt="Image" width="694" src="https://private-user-images.githubusercontent.com/30036188/408769276-152795e0-c095-424d-b847-bb353fb971b1.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5MzM5MDcsIm5iZiI6MTczODkzMzYwNywicGF0aCI6Ii8zMDAzNjE4OC80MDg3NjkyNzYtMTUyNzk1ZTAtYzA5NS00MjRkLWI4NDctYmIzNTNmYjk3MWIxLlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA3VDEzMDY0N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNkZGNhMDcwOTEwZWJmMzAzZjUyNmY3NGM5YWQwYzI0YmY1MDI1YmNmYjJiNzMwY2MwZDdmMDExODhlYTczNTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.pxqVkQgeNwcIM0NvjZBDQ8V9165LIDjLIqBLq-O5ung">

Hi
I'm also training this code. Currently, I have just generated the slice data of ShapeNetd, but there seem to be some issues during training. May I ask if you could share the training code for the regression slices? (This is my email: 202222000586@stu.swpu.edu.cn). Thank you very much if you can.

**Comment by yizhiwang96** [comment]
> Thank you for your reply. Is this the **CamerNet** that perform 3D to 2D projection? Is it to train this **CamerNet** and then add the --est_campose parameter in python reconstruct.py?
> 
> Could you please provide the **CamerNet** training code and let me know if there is anything else needed? Thank you!
> 
> ![Image](https://github.com/user-attachments/assets/e311dc3c-7ce4-4abd-a510-09343484ae07)


[**CamerNet**](https://github.com/yizhiwang96/Slice3D/blob/d02c0b307c94eb113feb3a5d4f6543cfdaad4dbb/reg_slices/src/model_cam_est.py#L13) is the network to estimate camera parameters from a single view. The training code is very similar to train.py. For now, can you use the GT camera parameters in reconstruction.py to see whether it works. I think it could involve some transformations between left-hand right-hand coordinate system but not sure. If  GT camera parameters work, you can go ahead to train a camera parameter estimation network.

**Comment by bellenliu** [comment]
> [**CamerNet**](https://github.com/yizhiwang96/Slice3D/blob/d02c0b307c94eb113feb3a5d4f6543cfdaad4dbb/reg_slices/src/model_cam_est.py#L13) is the network to estimate camera parameters from a single view. The training code is very similar to train.py. For now, can you use the GT camera parameters in reconstruction.py to see whether it works. I think it could involve some transformations between left-hand right-hand coordinate system but not sure. If GT camera parameters work, you can go ahead to train a camera parameter estimation network.

Thank you for your advice. I havd used the GT camera parameters in reconstruction.py.  I attached screenshot of my results reconstructed using epoch 80 ckpt. 

Could you please let me know is this result reasonable and which epoch of ckpt was used in the paper? Thank you!


![Image](https://github.com/user-attachments/assets/76cd5372-7692-4988-a5a4-b38e6e121a4c)

![Image](https://github.com/user-attachments/assets/294a5f6c-5efe-4852-9742-eec515884eac)

**Comment by yizhiwang96** [comment]
The best checkpoint was chosen based on the error/accuracy on the validation dataset. There is a train/val/test split for the ShapeNet.

The results look reasonable to me.

**Comment by ChongminWang** [comment]
![Image](https://github.com/user-attachments/assets/a8dde707-6e11-4cbe-821b-d1aaaddfb3b0)

![Image](https://github.com/user-attachments/assets/e9a58552-de38-4014-8872-81bde05ab6c0)

![Image](https://github.com/user-attachments/assets/54a9e849-00e0-49d0-8bdf-16059e8662a5)

![Image](https://github.com/user-attachments/assets/86a9cc2a-865f-45b8-9ccd-71a7328c9072) 
@yizhiwang96 I looked at the reconstruct.py code and found that both the ShapeNet data and the Objaverse data were used for GT camera parameters when testing. When I tested the Objaverse data, the results were basically consistent with the paper, but when I tested the ShapeNet data, the results did not look so good. The picture above is the result of my testing the chair category.  The picture above is the result of my chair testing. This is the model 560_721446_0.0121_0.8493_0.001637.ckpt  that I used for testing. In addition, I used ShapeNet data for training at the same time.

I have some questions?

(1) Is ShapeNet data trained separately in one category, or trained simultaneously in multiple categories?

(2) Can you provide an est_campose model of ShapeNet data that you have trained, such as [2023_1107_airplanes_est](https://github.com/yizhiwang96/Slice3D/blob/d02c0b307c94eb113feb3a5d4f6543cfdaad4dbb/reg_slices/options.py#L24C4-L24C17)

**Comment by yizhiwang96** [comment]
> ![Image](https://github.com/user-attachments/assets/a8dde707-6e11-4cbe-821b-d1aaaddfb3b0)
> 
> ![Image](https://github.com/user-attachments/assets/e9a58552-de38-4014-8872-81bde05ab6c0)
> 
> ![Image](https://github.com/user-attachments/assets/54a9e849-00e0-49d0-8bdf-16059e8662a5)
> 
> ![Image](https://github.com/user-attachments/assets/86a9cc2a-865f-45b8-9ccd-71a7328c9072) [@yizhiwang96](https://github.com/yizhiwang96) I looked at the reconstruct.py code and found that both the ShapeNet data and the Objaverse data were used for GT camera parameters when testing. When I tested the Objaverse data, the results were basically consistent with the paper, but when I tested the ShapeNet data, the results did not look so good. The picture above is the result of my testing the chair category. The picture above is the result of my chair testing. This is the model 560_721446_0.0121_0.8493_0.001637.ckpt that I used for testing. In addition, I used ShapeNet data for training at the same time.
> 
> I have some questions?
> 
> (1) Is ShapeNet data trained separately in one category, or trained simultaneously in multiple categories?
> 
> (2) Can you provide an est_campose model of ShapeNet data that you have trained, such as [2023_1107_airplanes_est](https://github.com/yizhiwang96/Slice3D/blob/d02c0b307c94eb113feb3a5d4f6543cfdaad4dbb/reg_slices/options.py#L24C4-L24C17)

Hi @ChongminWang , Objaverse and ShapeNet are quite different. The current code only supports Objaverse, which does not need any camera rotation parameters since it is camera-aligned slicing. **Based on your shown results, the camera projection is wrong.** 

Are you using the **axis-aligned** slicing to prepare the training data for ShapeNet?

If so, I think we should apply [estimated camera parameters](https://github.com/yizhiwang96/Slice3D/blob/main/reg_slices/reconstruct.py#L405) on this [step](https://github.com/yizhiwang96/Slice3D/blob/main/reg_slices/src/models.py#L53-L56) to align the 3D query points with 2D pixels for ShapeNet. Currently it is missing since camera-aligned slicing does not need it. Here estimated camera parameters can be replaced with GT ones to verify if the projection is correct. Can @bellenliu help to confirm if this is correct?

ShapeNet is trained and tested in each category separately.

**Comment by ChongminWang** [comment]
Hi @yizhiwang96 , I used the axis-aligned slicing to prepare the training data for ShapeNet, As you mentioned #[--slice_direction](https://github.com/yizhiwang96/Slice3D/issues/4#issuecomment-2430670867)
Maybe I really need to use the estimated camera parameters, but my ability is limited, do you still have the camera estimation model?
@bellenliu , did you solve the problem with the test?

**Comment by bellenliu** [comment]
> If so, I think we should apply [estimated camera parameters](https://github.com/yizhiwang96/Slice3D/blob/main/reg_slices/reconstruct.py#L405) on this [step](https://github.com/yizhiwang96/Slice3D/blob/main/reg_slices/src/models.py#L53-L56) to align the 3D query points with 2D pixels for ShapeNet. Currently it is missing since camera-aligned slicing does not need it. Here estimated camera parameters can be replaced with GT ones to verify if the projection is correct. Can [@bellenliu](https://github.com/bellenliu) help to confirm if this is correct?
> 
> ShapeNet is trained and tested in each category separately.


@ChongminWang Maybe you can change the data['qry_norot'] to fit the shapenet dataset. Hope it can help you! 
[Slice3D](https://github.com/yizhiwang96/Slice3D/tree/main)/[reg_slices](https://github.com/yizhiwang96/Slice3D/tree/main/reg_slices)
/reconstruct.py
> sdf_params = data['sdf_params'][0].cpu().numpy()
>  RESOLUTION = self.resolution0
> x_ = np.linspace(sdf_params[0], sdf_params[3], num=RESOLUTION)
> y_ = np.linspace(sdf_params[1], sdf_params[4], num=RESOLUTION)
> z_ = np.linspace(sdf_params[2], sdf_params[5], num=RESOLUTION)
>z, y, x = np.meshgrid(z_, y_, x_, indexing='ij')
>x = np.expand_dims(x, 3)
>y = np.expand_dims(y, 3)
>z = np.expand_dims(z, 3)
>all_pts = np.concatenate((x, y, z), axis=3).astype(np.float32)
> all_pts = all_pts.reshape( -1, 3)
> data['qry_norot'] = torch.from_numpy(all_pts).unsqueeze(0).cuda()

@yizhiwang96 Regarding the GT camera  parameters,  I use the GT camera parameters but get the coarse results. Maybe we need to change "img_w = 1" to "img_w =self.img_size"?   Could you please let me know if there is anything I might be doing wrong? Thank you!

[Slice3D](https://github.com/yizhiwang96/Slice3D/tree/main)/[reg_slices](https://github.com/yizhiwang96/Slice3D/tree/main/reg_slices)/[src](https://github.com/yizhiwang96/Slice3D/tree/main/reg_slices/src)
/datasets.py

> az = -data[1][cmr_angle_idx]
> el = data[2][cmr_angle_idx]
> distance = data[3][cmr_angle_idx]
> scale = data[5]
> offset = data[6]
> K, RT = getBlenderProj(az, el, distance, img_w=1, img_h=1)

![Image](https://github.com/user-attachments/assets/092b5aa5-43f1-4563-9343-ee1bf5555fef)

![Image](https://github.com/user-attachments/assets/ecc231dd-fcf1-4d77-8101-051ee64ca975)

**Comment by ChongminWang** [comment]
@bellenliu Thank you very much! I'll test it right away!

**Comment by ChongminWang** [comment]
> > If so, I think we should apply [estimated camera parameters](https://github.com/yizhiwang96/Slice3D/blob/main/reg_slices/reconstruct.py#L405) on this [step](https://github.com/yizhiwang96/Slice3D/blob/main/reg_slices/src/models.py#L53-L56) to align the 3D query points with 2D pixels for ShapeNet. Currently it is missing since camera-aligned slicing does not need it. Here estimated camera parameters can be replaced with GT ones to verify if the projection is correct. Can [@bellenliu](https://github.com/bellenliu) help to confirm if this is correct?
> > ShapeNet is trained and tested in each category separately.
> 
> [@ChongminWang](https://github.com/ChongminWang) Maybe you can change the data['qry_norot'] to fit the shapenet dataset. Hope it can help you! [Slice3D](https://github.com/yizhiwang96/Slice3D/tree/main)/[reg_slices](https://github.com/yizhiwang96/Slice3D/tree/main/reg_slices) /reconstruct.py
> 
> > sdf_params = data['sdf_params'][0].cpu().numpy()
> > RESOLUTION = self.resolution0
> > x_ = np.linspace(sdf_params[0], sdf_params[3], num=RESOLUTION)
> > y_ = np.linspace(sdf_params[1], sdf_params[4], num=RESOLUTION)
> > z_ = np.linspace(sdf_params[2], sdf_params[5], num=RESOLUTION)
> > z, y, x = np.meshgrid(z_, y_, x_, indexing='ij')
> > x = np.expand_dims(x, 3)
> > y = np.expand_dims(y, 3)
> > z = np.expand_dims(z, 3)
> > all_pts = np.concatenate((x, y, z), axis=3).astype(np.float32)
> > all_pts = all_pts.reshape( -1, 3)
> > data['qry_norot'] = torch.from_numpy(all_pts).unsqueeze(0).cuda()
> 
> [@yizhiwang96](https://github.com/yizhiwang96) Regarding the GT camera parameters, I use the GT camera parameters but get the coarse results. Maybe we need to change "img_w = 1" to "img_w =self.img_size"? Could you please let me know if there is anything I might be doing wrong? Thank you!
> 
> [Slice3D](https://github.com/yizhiwang96/Slice3D/tree/main)/[reg_slices](https://github.com/yizhiwang96/Slice3D/tree/main/reg_slices)/[src](https://github.com/yizhiwang96/Slice3D/tree/main/reg_slices/src) /datasets.py
> 
> > az = -data[1][cmr_angle_idx]
> > el = data[2][cmr_angle_idx]
> > distance = data[3][cmr_angle_idx]
> > scale = data[5]
> > offset = data[6]
> > K, RT = getBlenderProj(az, el, distance, img_w=1, img_h=1)
> 
> ![Image](https://github.com/user-attachments/assets/092b5aa5-43f1-4563-9343-ee1bf5555fef)
> 
> ![Image](https://github.com/user-attachments/assets/ecc231dd-fcf1-4d77-8101-051ee64ca975)

I have tried to use your method, but there has been a problem, the values of [points,  values](                [mesh_extractor.update(points, values)](https://github.com/yizhiwang96/Slice3D/blob/d02c0b307c94eb113feb3a5d4f6543cfdaad4dbb/reg_slices/reconstruct.py#L164)) do not match. Are you ready to share reconstruct.py files? 

![Image](https://github.com/user-attachments/assets/fe27b733-908c-4f53-a0db-4014d9eee02a)

By the way, I used the SDF file provided by [DISN](https://drive.google.com/file/d/1cHDickPLKLz3smQNpOGXD2W5mkXcy1nq/view).
When I use the reconstruct.py file provided by the author, the result of the reconstruction is always missing a corner, as in the example I gave:

![Image](https://github.com/user-attachments/assets/92d58cd0-a158-4a83-994c-3b26e1e6f1ec)
